File: 2023/ocaml/day01b/day01b.ml
open Printf

let usage () =
  let progname = Sys.argv.(0) in
  printf "usage: %s <file>\n" progname;
  exit 1

let rec find_str needle haystack at =
  if String.length haystack = 0 then
    -1
  else if String.starts_with ~prefix:needle haystack then
    at
  else
    find_str needle (String.sub haystack 1 ((String.length haystack) - 1)) (at + 1)

let rec rfind_str needle haystack at =
  if String.length haystack = 0 then
    -1
  else if String.ends_with ~suffix:needle haystack then
    at
  else
    rfind_str needle (String.sub haystack 0 ((String.length haystack) - 1)) (at - 1)

let process contents =
  let sum = ref 0 in
  let digits_map = Hashtbl.create 20 in
    Hashtbl.add digits_map "0" 0;
    Hashtbl.add digits_map "1" 1;
    Hashtbl.add digits_map "2" 2;
    Hashtbl.add digits_map "3" 3;
    Hashtbl.add digits_map "4" 4;
    Hashtbl.add digits_map "5" 5;
    Hashtbl.add digits_map "6" 6;
    Hashtbl.add digits_map "7" 7;
    Hashtbl.add digits_map "8" 8;
    Hashtbl.add digits_map "9" 9;
    Hashtbl.add digits_map "zero" 0;
    Hashtbl.add digits_map "one" 1;
    Hashtbl.add digits_map "two" 2;
    Hashtbl.add digits_map "three" 3;
    Hashtbl.add digits_map "four" 4;
    Hashtbl.add digits_map "five" 5;
    Hashtbl.add digits_map "six" 6;
    Hashtbl.add digits_map "seven" 7;
    Hashtbl.add digits_map "eight" 8;
    Hashtbl.add digits_map "nine" 9;
    contents |> String.split_on_char '\n' |> List.iter (fun line ->
      let min_index = ref None in
      let max_index = ref None in
      let left_digit = ref 0 in
      let right_digit = ref 0 in
      Hashtbl.iter (fun digit_str digit ->
        if (find_str digit_str line 0) <> -1 then
          let left_index = find_str digit_str line 0 in
          (match !min_index with
          | None -> min_index := Some left_index; left_digit := digit
          | Some mi when left_index < mi -> min_index := Some left_index; left_digit := digit
          | _ -> ());
          let right_index = rfind_str digit_str line ((String.length line) - 1) in
          (match !max_index with
          | None -> max_index := Some right_index; right_digit := digit
          | Some mi when right_index > mi -> max_index := Some right_index; right_digit := digit
          | _ -> ());
      ) digits_map;
      sum := !sum + (!left_digit * 10) + !right_digit
    );
    !sum

let () =
  if Array.length Sys.argv < 2 then
    usage ()
  else
    let filename = Sys.argv.(1) in
    let input_channel = open_in filename in
    try
      let contents = In_channel.input_all input_channel in
      let result = process contents in
        printf "result = %d\n" result
    with e ->
      raise e

File: 2023/r/day01b/day01b.r
#!/usr/bin/env Rscript

prog_name <- "day01b.r"

usage <- function() {
    cat("usage:", prog_name, "<file>\n")
    quit(status=1)
}

process <- function(filename) {
    sum <- 0
    digit_strs <- list(
        "0" = 0,
        "1" = 1,
        "2" = 2,
        "3" = 3,
        "4" = 4,
        "5" = 5,
        "6" = 6,
        "7" = 7,
        "8" = 8,
        "9" = 9,
        "zero" = 0,
        "one" = 1,
        "two" = 2,
        "three" = 3,
        "four" = 4,
        "five" = 5,
        "six" = 6,
        "seven" = 7,
        "eight" = 8,
        "nine" = 9
    )
    infile <- file(filename, "r")
    while ( length(line <- readLines(infile, 1)) > 0 ) {
        min_index <- NULL
        max_index <- NULL
        left_digit <- 0
        right_digit <- 0
        for ( digit_str in names(digit_strs) ) {
            if ( grepl(digit_str, line) ) {
                left_index <- min(gregexpr(digit_str, line)[[1]])
                if ( is.null(min_index) || left_index < min_index ) {
                    min_index <- left_index
                    left_digit <- digit_strs[[digit_str]]
                }
                right_index <- max(gregexpr(digit_str, line)[[1]])
                if ( is.null(max_index) || right_index > max_index ) {
                    max_index <- right_index
                    right_digit <- digit_strs[[digit_str]]
                }
            }
        }
        sum <- sum + (left_digit * 10) + right_digit
    }
    sum
}

main <- function() {
    args <- commandArgs(trailingOnly=TRUE)
    if ( length(args) < 1 ) {
        usage()
    }
    filename <- args[1]
    result <- process(filename)
    cat("result =", result, "\n")
}

if (!interactive()) {
    main()
}
